File: app/src/main/java/com/noobfitness/noobfitness/auth/LoginActivity.java
package com.noobfitness.noobfitness.auth;

import android.app.Activity;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import com.google.android.gms.common.SignInButton;
import com.noobfitness.noobfitness.MainApplication;
import com.noobfitness.noobfitness.R;
import com.noobfitness.noobfitness.legacy.MainActivity;

import net.openid.appauth.AuthState;
import net.openid.appauth.AuthorizationRequest;
import net.openid.appauth.AuthorizationService;
import net.openid.appauth.AuthorizationServiceConfiguration;

public class LoginActivity extends Activity {

    private static final String LOG_TAG = "LoginActivity";

    MainApplication mMainApplication;

    // views
    SignInButton mAuthorize;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        mMainApplication = (MainApplication) getApplication();

        mAuthorize = findViewById(R.id.authorize);

        // wire click listeners
        mAuthorize.setOnClickListener(new AuthorizeListener());
    }

    /**
     * Kicks off the authorization flow.
     */
    public static class AuthorizeListener implements Button.OnClickListener {
        @Override
        public void onClick(View view) {
            Context context = view.getContext();

            // code from the step 'Create the Authorization Request',
            // and the step 'Perform the Authorization Request' goes here.
            AuthorizationServiceConfiguration serviceConfiguration = new AuthorizationServiceConfiguration(
                    Uri.parse("https://accounts.google.com/o/oauth2/v2/auth") /* auth endpoint */,
                    Uri.parse("https://www.googleapis.com/oauth2/v4/token") /* token endpoint */
            );

            String clientId = context.getString(R.string.google_client_id);
            Uri redirectUri = Uri.parse("com.noobfitness.noobfitness:/oauth2callback");
            AuthorizationRequest.Builder builder = new AuthorizationRequest.Builder(
                    serviceConfiguration,
                    clientId,
                    "code",
                    redirectUri
            );
            builder.setScopes("openid profile email");
            AuthorizationRequest request = builder.build();

            AuthorizationService authorizationService = new AuthorizationService(context);

            Intent postAuthorizationIntent = new Intent(context, MainActivity.class);
            PendingIntent pendingIntent = PendingIntent.getActivity(context, request.hashCode(), postAuthorizationIntent, 0);
            authorizationService.performAuthorizationRequest(request, pendingIntent);
        }
    }
}

File: app/src/main/java/com/noobfitness/noobfitness/auth/LoginActivity.kt
package com.noobfitness.noobfitness.auth

import android.app.Activity
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle

import com.google.android.gms.common.SignInButton
import com.noobfitness.noobfitness.dagger.InjectedApplication
import com.noobfitness.noobfitness.R
import com.noobfitness.noobfitness.legacy.MainActivity
import net.openid.appauth.*
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject

import javax.inject.Inject

class LoginActivity : Activity() {

    @Inject
    lateinit var authService: AuthService
    @Inject
    lateinit var userManager: UserManager

    private lateinit var button: SignInButton
    private var authState: AuthState? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (application as InjectedApplication).appComponent.inject(this)

        setContentView(R.layout.activity_login)

        button = findViewById(R.id.sign_in_button)
        button.setOnClickListener {
            val intent = authService.getAuthorizationRequestIntent()
            startActivityForResult(intent, RC_AUTH)
        }

        if (userManager.get() != null) {
            login()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == RC_AUTH) {
            val response = AuthorizationResponse.fromIntent(data!!)
            val error = AuthorizationException.fromIntent(data!!)

            onAuthorizationResponse(response, error)
        }
    }

    private fun onAuthorizationResponse(response: AuthorizationResponse?, error: AuthorizationException?) {
        if (response != null) {
            authState = AuthState(response, error)
            authService.performTokenRequest(response.createTokenExchangeRequest(), ::onGAuthTokenResponse)
        }
    }

    private fun onGAuthTokenResponse(response: TokenResponse?, error: AuthorizationException?) {
        if (response != null) {
            authState?.apply { update(response, error) }

            object : AsyncTask<String, Void, User>() {
                override fun doInBackground(vararg tokens: String): User? {
                    val client = OkHttpClient()
                    val requestBody = FormBody.Builder()
                            .add("access_token", tokens[0])
                            .build()
                    val request = Request.Builder()
                            .post(requestBody)
                            .url("http://10.0.2.2:5000/api/auth/google")
                            .build()
                    val response = client.newCall(request).execute()
                    val authToken = response.header("x-auth-token")!!
                    val jsonBody = response.body()!!.string()
                    val userJson = JSONObject(jsonBody)

                    val id = userJson.optString("_id", null)
                    val googleId = userJson.optString("googleId", null)
                    val username = userJson.optString("username", null)
                    val avatar = userJson.optString("avatar", null)
                    val routines = userJson.optString("routines", null)

                    return User(
                            authToken = authToken,
                            id = id,
                            googleId = googleId,
                            username = username,
                            avatar = avatar,
                            routines = routines
                    )
                }

                override fun onPostExecute(user: User?) {
                    if (user != null) {
                        userManager.set(user)
                        login()
                    }
                }
            }.execute(authState?.accessToken)
        }
    }

    private fun login() {
        button.isEnabled = false

        val intent = Intent(this, MainActivity::class.java).apply {
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
        }

        startActivity(intent)
    }

    companion object {
        private const val RC_AUTH = 100
    }
}
